src/ConfigFileParser.cpp

```diff
@@ -0,0 +1,534 @@
+        port = num;
+    } else
+        port = value.substr(colonIndex + 1, semicolonIndex - colonIndex - 1);
+
+    if (!isPortValid(port))
+        throw std::runtime_error("Invalid port");
+
+    for (std::map<std::string, unsigned short>::iterator it = m_configFile.servers[m_configFile.serverIndex].listen.begin(); it != m_configFile.servers[m_configFile.serverIndex].listen.end(); it++)
+        if (it->second == 0)
+            it->second = std::atoi(port.c_str());
+}
+
+/**
+ * @brief Reads the root path
+ *
+ * The function checks if the root path is valid and reads it if that is the case.
+ *
+ * It makes sure that the path is valid in the following ways:
+ * 1. There is only one root path
+ *
+ * If at the end of the path is a slash, it removes it.
+ *
+ * This function can be used for the server block and location block
+ *
+ * @param block The block which surounds the directive
+ */
+void ConfigFileParser::readRootPath(int block, const std::string& value)
+{
+    std::string rootPath = value;
+
+    if (rootPath.find_first_of("  \t\n\v\f\r") != std::string::npos)
+        throw std::runtime_error("More than one root path");
+
+    if (rootPath[rootPath.length() - 1] == '/')
+        rootPath = rootPath.substr(0, rootPath.length() - 1);
+
+    if (block == SERVER)
+        m_configFile.servers[m_configFile.serverIndex].root = rootPath;
+    else if (block == LOCATION)
+        m_configFile.servers[m_configFile.serverIndex].locations[m_configFile.servers[m_configFile.serverIndex].locationIndex].root = rootPath;
+}
+
+/**
+ * @brief Reads and checks the value of the directive in the current line of the config file
+ *
+ * @details This function is called when the directive is valid.
+ *          It calls the appropriate function to read the value of the directive.
+ *          It throws an exception if the value is invalid.
+
+ * @param directive Is the the directive which value is being read and checked
+ */
+void ConfigFileParser::readServerDirectiveValue(const std::string& directive, const std::string& value)
+{
+    if (directive == "listen") {
+        readIpAddress(value);
+        readPort(value);
+    } else if (directive == "root")
+        readRootPath(SERVER, value);
+}
+
+/**
+ * @brief Gets the directive from the line
+ *
+ * @param line String containing the directive and value
+ * @return std::string The extracted directive
+ */
+std::string ConfigFileParser::getDirective(const std::string& line) const
+{
+    size_t firstWhiteSpaceIndex = line.find_first_of(" \t\n\v\f\r");
+    std::string directive = line.substr(0, firstWhiteSpaceIndex);
+
+    removeLeadingAndTrailingSpaces(directive);
+
+    return directive;
+}
+
+/**
+ * @brief Gets the value from the line
+ *
+ * @param line String containing the directive and value;
+ * @return std::string  The extracted value
+ */
+std::string ConfigFileParser::getValue(const std::string& line) const
+{
+    if (getDirective(line) == "location")
+        return line;
```

Comment:
I think we can return "" (empty string) in this case, since we won't check value anyway.

I would even propose to move this check out of this function and check before going in so you don't need to extract the directive twice